File: CT/pubsub.js
CT.pubsub = {
	"_": {
		"ws": null,
		"open": false,
		"reconnect": true,
		"reconnect_interval": 250,
		"initialized": false,
		"queue": [],
		"channels": {},
		"cb": {
			"message": CT.data.getLogger("CT.pubsub|message"), // override w/ set_cb()
			"subscribe": CT.data.getLogger("CT.pubsub|subscribe"), // override w/ set_cb()
			"join": CT.data.getLogger("CT.pubsub|join"), // override w/ set_cb()
			"leave": CT.data.getLogger("CT.pubsub|leave"), // override w/ set_cb()
			"open": CT.data.getLogger("CT.pubsub|open"), // override w/ set_cb()
			"close": CT.data.getLogger("CT.pubsub|close"), // override w/ set_cb()
			"error": CT.data.getLogger("CT.pubsub|error"), // override w/ set_cb()
		},
		"process": {
			"read": function(msg) {
				var d = JSON.parse(msg.data);
				CT.pubsub._.process[d.action](d.data);
			},
			"channel": function(data) {
				CT.pubsub._channels[data.channel] = data;
				data.history.forEach(CT.pubsub._.process.publish);
				CT.pubsub._.process.subscribe(data.channel, data);
			},
			"publish": function(data) {
				CT.pubsub._.cb.message(data);
			},
			"subscribe": function(data) {
				CT.pubsub._.cb.join(data.channel, data.user);
				CT.data.add(CT.pubsub._.channels[data.channel].users, data.user);
			},
			"unsubscribe": function(data) {
				CT.pubsub._.cb.leave(data.channel, data.user);
				CT.data.remove(CT.pubsub._.channels[data.channel].users, data.user);
			}
		}
	},
	"_register": function() {
		CT.pubsub._.open = true;
		CT.pubsub._.queue.forEach(function(item, i) {
			setTimeout(function() {
				CT.pubsub.write(item);
			}, 100 * i);
		});
		CT.pubsub._.cb.open("opened");
	},
	"isInitialized": function() {
		return CT.pubsub._.initialized;
	},
	"set_reconnect": function(bool) {
		CT.pubsub._.reconnect = bool;
	},
	"set_cb": function(action, cb) { // action: message|join|leave|open|close|error
		CT.pubsub.cb[action] = cb;
	},
	"write": function(data) {
		if (CT.pubsub._.open) {
			var dstring = JSON.stringify(data);
			CT.pubsub._.ws.send(dstring);
			console.log("WRITE", dstring);
		} else {
			CT.pubsub._.queue.push(data);
			console.log("QUEUE", JSON.stringify(CT.pubsub._.queue));
		}
	},
	"publish": function(channel, message) {
		CT.pubsub.write({
			"action": "publish",
			"data": {
				"channel": channel,
				"message": message
			}
		});
	},
	"subscribe": function(channel) {
		CT.pubsub.write({
			"action": "subscribe",
			"data": channel
		});
	},
	"unsubscribe": function(channel) {
		CT.pubsub.write({
			"action": "unsubscribe",
			"data": channel
		});
	},
	"connect": function(host, port, uname) {
		CT.pubsub._.initialized = true;
		CT.pubsub._.ws = new WebSocket("ws://" + host + ":" + port);
		CT.pubsub._.ws.onopen = CT.pubsub._register;
		CT.pubsub._.ws.onmessage = CT.pubsub._read;
		CT.pubsub._.ws.onclose = function() {
			CT.pubsub._.cb.close("closed");
		};
		CT.pubsub._.ws.onerror = function(e) {
			CT.pubsub._.cb.error("error", e);
		};
		CT.pubsub.write(uname);
	}
};

File: scripts/pubsub.py
import json
from dez.network.websocket import WebSocketDaemon
from util import log
from config import PUBSUB_PORT, PUBSUB_HIST

class PubSub(WebSocketDaemon):
    def __init__(self, *args, **kwargs):
        kwargs["isJSON"] = True
        kwargs["report_cb"] = self._log
        kwargs["cb"] = self.connect
        if "silent" in kwargs:
            self.silent = kwargs["silent"]
            del kwargs["silent"]
        else:
            self.silent = False
        WebSocketDaemon.__init__(self, *args, **kwargs)
        self.clients = {}
        self.channels = {}
        log("Initialized PubSub Server @ %s:%s"%(self.hostname, self.port), important=True)

    def subscribe(self, channel, user):
        self._check_channel(channel)
        chan = self.channels[channel]
        chan.join(user)
        self._log('SUBSCRIBE: "%s" -> "%s"'%(user.name, channel), 2)
        user.write({
            "action": "channel",
            "data": {
                "channel": channel,
                "users": [u.name for u in chan.users],
                "history": chan.history
            }
        })

    def unsubscribe(self, channel, user):
        if self._check_channel(channel, True) and user in self.channels[channel].users:
            self.channels[channel].users.remove(user)
            self._log('UNSUBSCRIBE: "%s" -> "%s"'%(user.name, channel), 2)
        else:
            self._log('FAILED UNSUBSCRIBE: "%s" -> "%s"'%(user.name, channel), 2)

    def publish(self, data, user):
        channel = data["channel"]
        self._check_channel(channel)
        self.channels[channel].write({
            "message": data["message"],
            "user": user.name
        })

    def _check_channel(self, channel, justBool=False):
        condition = channel in self.channels
        if not condition and not justBool:
            self.channels[channel] = PubSubChannel(channel, self._log)
        return condition

    def _log(self, data, level=0, important=False):
        if not self.silent:
            log(data, level=level, important=important)

    def connect(self, conn):
        PubSubUser(conn, self, self._log)

class PubSubChannel(object):
    def __init__(self, name, logger):
        self._log = logger
        self.name = name
        self.users = set()
        self.history = []
        self._log('NEW CHANNEL: "%s"'%(name,), 1, True)

    def write(self, obj):
        obj["channel"] = self.name
        dstring = json.dumps({
            "action": "publish",
            "data": obj
        })
        for user in self.users:
            user.conn.write(dstring, noEncode=True) # skips logging
        self.history.append(obj)
        self.history = self.history[:PUBSUB_HIST]

    def join(self, user):
        dstring = json.dumps({
            "action": "subscribe",
            "data": {
                "user": user.name,
                "channel": self.name
            }
        })
        for user in self.users:
            user.conn.write(dstring, noEncode=True) # skips logging
        self.users.add(user)

class PubSubUser(object):
    def __init__(self, conn, server, logger):
        self.conn = conn
        self.server = server
        self._log = logger
        self.conn.set_cb(self._register)
        self._log('NEW CONNECTION', 1, True)

    def write(self, data):
        dstring = json.dumps(data) # pre-encode so we can log
        self._log('WRITE: "%s" -> "%s"'%(self.name, dstring), 3)
        self.conn.write(dstring, noEncode=True)

    def _read(self, obj):
        getattr(self.server, obj["action"])(obj["data"], self)

    def _register(self, name):
        self._log('REGISTER: "%s"'%(name,), 1, True)
        self.name = name
        self.server.clients[name] = self
        self.conn.set_cb(self._read)

if __name__ == "__main__":
    PubSub('localhost', PUBSUB_PORT).start()